api/server/flexvolume.go

```diff
@@ -1,27 +1,155 @@
+	// ErrMissingMountPath is returned when no mountPath specified in mount call
+	ErrMissingMountPath = errors.New("Missing target mount path")
+)
+
+type flexVolumeClient struct{}
 
 func (c *flexVolumeClient) Init() error {
 	return nil
 }
 
 func (c *flexVolumeClient) Attach(jsonOptions map[string]string) error {
+	driverName, ok := jsonOptions[osdDriverKey]
+	if !ok {
+		return ErrInvalidSpecDriver
+	}
+
+	driver, err := c.getVolumeDriver(driverName)
+	if err != nil {
+		return err
+	}
+
+	mountDevice, ok := jsonOptions[volumeIDKey]
+	if !ok {
+		return ErrInvalidSpecVolumeID
+	}
+
+	if _, err = driver.Attach(mountDevice); err != nil {
+		return err
+	}
 	return nil
 }
 
 func (c *flexVolumeClient) Detach(mountDevice string) error {
+	driverName, ok := deviceDriverMap[mountDevice]
+	if !ok {
+		dlog.Infof("Could not find driver for (%v). Defaulting to pxd", mountDevice)
+		driverName = pxDriverKey
+	}
+
+	driver, err := c.getVolumeDriver(driverName)
+	if err != nil {
+		return err
+	}
+
+	if err = driver.Detach(mountDevice); err != nil {
+		return err
+	}
 	return nil
 }
 
-func (c *flexVolumeClient) Mount(targetMountDir string, mountDevice string, jsonOptions map[string]string) error {
+func (c *flexVolumeClient) Mount(targetMountDir string, mountDevice string,
+	jsonOptions map[string]string) error {
+
+	driverName, ok := jsonOptions[osdDriverKey]
+	if !ok {
+		return ErrInvalidSpecDriver
+	}
+
+	driver, err := c.getVolumeDriver(driverName)
+	if err != nil {
+		return err
+	}
+
+	if targetMountDir == "" {
+		return ErrMissingMountPath
+	}
+
+	if err = driver.Mount(mountDevice, targetMountDir); err != nil {
+		return err
+	}
+
+	// Update the deviceDriverMap
+	mountManager, err := mount.New(mount.DeviceMount, "")
+	if err != nil {
+		dlog.Infof("Could not read mountpoints from /proc/self/mountinfo. Device - Driver mapping not saved!")
+		return nil
+	}
+
+	devPath, err := mountManager.GetDevPath(targetMountDir)
+	if err != nil {
+		return fmt.Errorf("Mount Failed. Could not find mountpoint in /proc/self/mountinfo")
+	}
+	deviceDriverMap[devPath] = driverName
 	return nil
 }
 
 func (c *flexVolumeClient) Unmount(mountDir string) error {
+	var driverName string
+
+	// Get the mountDevice from mount manager
+	mountManager, err := mount.New(mount.DeviceMount, "")
+	if err != nil {
+		return ErrNoMountInfo
+	}
+
+	mountDevice, err := mountManager.GetDevPath(mountDir)
+	if err != nil {
+		return fmt.Errorf("Invalid mountDir (%v). No device mounted on this path", mountDir)
+	}
+
+	driverName, ok := deviceDriverMap[mountDevice]
+	if !ok {
+		dlog.Infof("Could not find driver for (%v). Defaulting to pxd", mountDevice)
+		driverName = pxDriverKey
+	}
+
+	driver, err := c.getVolumeDriver(driverName)
+	if err != nil {
+		return err
+	}
+
+	if err = driver.Unmount(mountDevice, mountDir); err != nil {
+		return err
+	}
 	return nil
 }
+
+func newFlexVolumeClient() *flexVolumeClient {
+	deviceDriverMap = make(map[string]string)
```

Comment:
setting a global variable in a constructor? why not just doing this when you declare deviceDriverMap?